### PR TITLE
Sporadic exit exception

### LINKED ISSUE #864: Sporadic exception upon application exit
An sporadic exception is occurring upon system exit, and it seems to be related to daemon threads throwing exceptions when resources are pulled out from underneath is - even if they're set to daemon.

this is showing up at a high rate in Sentry. 

This conversation seems to be the most likely explanation:
http://stackoverflow.com/questions/4079810/python-thread-exception-errors-when-exit-the-whole-program

-------------
Exception:

'NoneType' object is not subscriptable
home/brent/git-projects/RaceCapture_App/.buildozer/android/platform/python-for-android/build/python-install/lib/python2.7/site-packages/kivy/core/window/window_pygame.py in mainloop at line 398

home/brent/git-projects/RaceCapture_App/.buildozer/android/platform/python-for-android/build/python-install/lib/python2.7/site-packages/kivy/core/window/window_pygame.py in _mainloop at line 285

home/brent/git-projects/RaceCapture_App/.buildozer/android/platform/python-for-android/build/python-install/lib/python2.7/site-packages/kivy/base.py in idle at line 333

home/brent/git-projects/RaceCapture_App/.buildozer/android/platform/python-for-android/build/python-install/lib/python2.7/site-packages/kivy/lang.py in sync at line 2082

### PR DESCRIPTION
**Pull request by brentpicasso**
I think this resolves #864.  I was able to repro the exception upon manual exception (seemed to happen 1 out of 10 times) 

This patch has the msg_rx_worker exit when the app exits. pretty simple. After this patch I wasn't able to repro the exit exception after about 30-40 attempts.   The real tell is if the Sentry exceptions will go away or reduce dramatically.

### CODE REVIEW

File: autosportlabs/racecapture/api/rcpapi.py [hunk 1]
```diff
@@ -114,6 +114,7 @@ def init_comms(self, comms):
 
     def cleanup_comms(self):
```
Comment (stieg) [inline autosportlabs/racecapture/api/rcpapi.py]: So what is the difference between `cleanup_comms` and `shutdown_comms` exactly?  I'm just trying to figure out if line 117 is the correct place to put this?  The location is confusing since the order in which these methods should be called isn't entirely clear.  I assume it would go `shutdown_comms` then `cleanup_comms` but that is a guess.  Will you clarify that please/

Comment (brentpicasso) [inline autosportlabs/racecapture/api/rcpapi.py]: shutdown_comms is only used by the firmware update screen and only shuts down the rc_api message loop, so it won't try to grab and process data while the firmware is updating.   There are other daemon threads that remain running in the background. 

shutdown_comms could be renamed something more accurate. will investigate.

Comment (brentpicasso) [inline autosportlabs/racecapture/api/rcpapi.py]: Well, further digging, the function name is technically correct. The firmware update screen disables auto-connection recovery, *then* shuts down any comm port activity. The rc_api idles until the firmware update screen re-enables it. The message loop thread *does* remain alive, it just pauses waiting for data.

A better set of functions might be "pause_api" and "resume_api".   Also place the other daemon threads under control so they can be explicitly exited like rc_api thread.

Comment (stieg) [inline autosportlabs/racecapture/api/rcpapi.py]: Sounds like a state machine implementation there.  So yeah those method names sound much more descriptive.

File: autosportlabs/racecapture/api/rcpapi.py [hunk 2]
```diff
@@ -91,29 +91,38 @@ def recover_connection(self):
 
     def _start_message_rx_worker(self):
         self._running.set()
-        self._msg_rx_thread = Thread(target=self.msg_rx_worker)
-        self._msg_rx_thread.daemon = True
-        self._msg_rx_thread.start()
+        t = Thread(target=self.msg_rx_worker)
+        t.daemon = True
```
Comment (stieg) [inline autosportlabs/racecapture/api/rcpapi.py]: If you are having this issue both the [Referenced Stackoverflow Question](http://stackoverflow.com/questions/4079810/python-thread-exception-errors-when-exit-the-whole-program) and [Python threading Docs](https://docs.python.org/2/library/threading.html) mention not to set the Daemonic flag.  Is there a reason you are still setting it?

Comment (brentpicasso) [inline autosportlabs/racecapture/api/rcpapi.py]: I left it enabled b/c harmless and it's a backup in case we've failed to terminate it cleanly, somehow. As a daemon thread the python runtime will terminate it regardless.

File: autosportlabs/racecapture/api/rcpapi.py [hunk 3]
```diff
@@ -91,29 +91,38 @@ def recover_connection(self):
...
+        t.start()
+        self._msg_rx_thread = t
 
-    def stop_message_rx_worker(self):
+    def _shutdown_workers(self):
         Logger.info('RCPAPI: Stopping msg rx worker')
         self._running.clear()
+        # this allows the auto detect worker to fall through if needed
+        self._auto_detect_event.set()
```
Comment (stieg) [inline autosportlabs/racecapture/api/rcpapi.py]: Would it be wiser to set these variables _after_ you have joined the threads?

File: autosportlabs/racecapture/api/rcpapi.py [hunk 4]
```diff
@@ -91,29 +91,38 @@ def recover_connection(self):
...
+    def init_api(self, comms):
         self.comms = comms
         self._start_message_rx_worker()
         self._start_cmd_sequence_worker()
         self.start_auto_detect_worker()
         Clock.schedule_interval(lambda dt: comms.keep_alive(), COMMS_KEEP_ALIVE_TIMEOUT)
 
-    def cleanup_comms(self):
-        self.comms.cleanup()
+    def shutdown_api(self):
```
Comment (stieg) [inline autosportlabs/racecapture/api/rcpapi.py]: Nice to see things getting more consistent :)

File: autosportlabs/racecapture/api/rcpapi.py (hunk 3 above)
Comment (brentpicasso) [inline autosportlabs/racecapture/api/rcpapi.py]: It's specifically this way - it triggers an auto-detect so it can fall through and be caught, and then exited. It's a bit of a hack, but works.

File: autosportlabs/racecapture/api/rcpapi.py [hunk 5]
```diff
@@ -223,9 +232,11 @@ def _queue_multiple(self, command_list, root_name, win_callback, fail_callback):
         self._command_queue.put(command)
 
     def cmd_sequence_worker(self):
-        while True:
+        Logger.info('RCPAPI: cmd_sequence_worker starting')
+        while self._running.is_set():
             try:
-                command = self._command_queue.get() #this blocks forever
+                # Block for 1 second for messages
+                command = self._command_queue.get(True, 1.0)
```
Comment (stieg) [inline autosportlabs/racecapture/api/rcpapi.py]: Could set this to something smaller like 0.5 to make quit more responsive.  Perhaps define a magic value at the top and use it for consistency?

Comment (brentpicasso) [inline autosportlabs/racecapture/api/rcpapi.py]: Yes - need to fix the magic number, but it's only needed when exiting.

File: autosportlabs/racecapture/api/rcpapi.py [hunk 6]
```diff
@@ -625,16 +645,21 @@ def on_ver_win(value):
...
-        while True:
+        while self._running.is_set():
             try:
                 self._auto_detect_event.wait()
                 self._auto_detect_event.clear()
                 self._enable_autodetect.wait()
+                # check again if we're shutting down
+                # to prevent a needless re-detection attempt
+                if not self._running.is_set():
+                    break
```
Comment (stieg) [inline autosportlabs/racecapture/api/rcpapi.py]: Seems strange to have a duplicate check to the while here.  Would re-arranging the order fix this?

File: autosportlabs/racecapture/api/rcpapi.py [hunk 7]
```diff
@@ -690,8 +715,11 @@ def on_ver_win(value):
                 Logger.error('RCPAPI: Error running auto detect: ' + str(e))
                 Logger.debug(traceback.format_exc())
             finally:
-                Logger.info("RCPAPI: auto detect finished. port=" + str(comms.port))
-                self._auto_detect_busy.clear()
-                self.removeListener("ver", on_ver_win)
-                self.sendCommandLock.release()
-                sleep(AUTODETECT_COOLOFF_TIME)
+                if self._running.is_set():
```
Comment (stieg) [inline autosportlabs/racecapture/api/rcpapi.py]: Why is this needed?  Why can't the program just cleanup like it supposed to?

File: autosportlabs/racecapture/databus/databus.py [hunk 8]
```diff
@@ -193,6 +192,13 @@ def startDataPump(self, data_bus, rc_api):
             # we're already running, refresh channel meta data
             self.meta_is_stale()
 
+    def stop(self):
+        self._running.clear()
+        try:
+            self._sample_thread.join()
```
Comment (stieg) [inline autosportlabs/racecapture/databus/databus.py]: Strange that this would fail.  Any reason?

File: autosportlabs/racecapture/api/rcpapi.py (hunk 2 above)
Comment (stieg) [inline autosportlabs/racecapture/api/rcpapi.py]: I would argue that if we are failing to terminate cleanly we should fail and fix it.  Its a programming error at that point.  I argue we should nuke this flag and fix it properly so we exit like a well behaved program.

File: autosportlabs/racecapture/api/rcpapi.py (hunk 6 above)
Comment (brentpicasso) [inline autosportlabs/racecapture/api/rcpapi.py]: it's to cause the auto-detection to fall through immediately if we're self._running is cleared (signal to not run anymore).

Otherwise it'll needlessly trigger a re-detection effort, which will cause noise and delays in the app exiting.

File: autosportlabs/racecapture/api/rcpapi.py (hunk 7 above)
Comment (brentpicasso) [inline autosportlabs/racecapture/api/rcpapi.py]: It was needed b/c when it would fall through due to the break it would leave things in an unexpected state.  However, I moved the try block down after the break so I wouldn't need to deal with this situation; so, it's cleaner.

File: autosportlabs/racecapture/api/rcpapi.py (hunk 2 above)
Comment (brentpicasso) [inline autosportlabs/racecapture/api/rcpapi.py]: There is a small chance where there's a race condition when we shut down the app we try to join a thread that doesn't exist yet, *then* it creates the thread immediately afterwards. 

If the thread isn't daemonic, there would be left zombie threads lying around in people's task manager.  The daemon thread is a backup to handle this rare case.

File: autosportlabs/racecapture/databus/databus.py (hunk 8 above)
Comment (brentpicasso) [inline autosportlabs/racecapture/databus/databus.py]: Mostly in case we try to exit *before* the thread is created. 

It's more pythonic to ask for forgiveness than to check if something is possible.  See http://stackoverflow.com/questions/7604636/better-to-try-something-and-catch-the-exception-or-test-if-its-possible-first

File: autosportlabs/racecapture/api/rcpapi.py (hunk 2 above)
Comment (stieg) [inline autosportlabs/racecapture/api/rcpapi.py]: Huh?  I figure it would be that case if it _was_ a deamonic thread.  If it was a child thread without a parent then init would reap it unless it was treated as a deamon.

Comment (brentpicasso) [inline autosportlabs/racecapture/api/rcpapi.py]: Back to the original issue - daemon threads were fine, but they were sometimes generating otherwise harmless exceptions upon application shut down. These shut-down exceptions were showing up on Sentry in a high volume. 

So this is all in an effort to get threads to shut down cleanly, to minimize those harmless exceptions.   I had thought of squelching those exceptions in the crash handler, but thought attempting to close them down cleanly first would be the better design.

Comment (stieg) [inline autosportlabs/racecapture/api/rcpapi.py]: > I had thought of squelching those exceptions in the crash handler, but thought attempting to close them down cleanly first would be the better design.

I agree.  This is the proper approach.  So here is the line that is concerning to me from the Python docs:

> Daemon threads are abruptly stopped at shutdown. Their resources (such as open files, database transactions, etc.) may not be released properly. If you want your threads to stop gracefully, make them non-daemonic and use a suitable signalling mechanism such as an Event.

This is why I am persistent about this.  Because this very sentence seems to counter what you are trying to do.  Perhaps I am mistaken.

Comment (brentpicasso) [inline autosportlabs/racecapture/api/rcpapi.py]: We are attempting to shut them down before python shuts down at the very end - called from the on_stop() in main.py 

If for some reason they don't get stopped cleanly, the python runtime will come through later, at the very end and do a final cleanup. 

If they are abruptly stopped there may be a chance that a resource the thread was dependent upon would go away, causing those harmless but noisy exceptions to show up in the Sentry logs.
